nn_adapt/solving_one2n.py

```diff
@@ -0,0 +1,244 @@
+        q_plus = Function(V)
+        solver_obj_plus = config.Solver_one2n(refined_mesh, q_plus, **kwargs)
+        solver_obj_plus.iterate()
+        q_plus = solver_obj_plus.solution
+        # J = config.get_qoi(refined_mesh[-1])(q_plus[-1])
+        adj_solution_plus = []
+        solver_obj_plus.adjoint_iteration()
+        adj_solution_plus = solver_obj_plus.adj_solution
+        out["enriched_adjoint"] = adj_solution_plus
+        
+    out["times"]["estimator"] += perf_counter()
+    
+    return out
+
+
+def split_into_components(f):
+    r"""
+    Extend the :attr:`split` method to apply
+    to non-mixed :class:`Function`\s.
+    """
+    return [f] if f.function_space().value_size == 1 else f.split()
+
+
+def indicate_errors_one2n(mesh, config, enrichment_method="h", retall=False, **kwargs):
+    """
+    Indicate errors according to ``dwr_indicator``,
+    using the solver given in the configuration file.
+
+    :arg mesh: input mesh
+    :arg config: configuration file, which
+        specifies the PDE and QoI
+    :kwarg enrichment_method: how to enrich the
+        finite element space?
+    :kwarg retall: if ``True``, return the forward
+        solution and adjoint solution in addition
+        to the dual-weighted residual error indicator
+    """
+    if not enrichment_method == "h":
+        raise NotImplementedError  # TODO
+    with PETSc.Log.Event("Enrichment"):
+        tt_steps = config.parameters.tt_steps
+        mesh, ref_mesh = MeshHierarchy(mesh, 1)
+
+    # Solve the forward and adjoint problems
+    out = get_solutions_one2n(mesh=mesh, config=config, refined_mesh=ref_mesh, **kwargs)
+    if retall and "adjoint" not in out:
+        return out
+
+    out["times"]["estimator"] -= perf_counter()
+    
+    with PETSc.Log.Event("Enrichment"):
+        adj_sol_plus = out["enriched_adjoint"]
+        dwr = 0
+        dwr_list = []
+        
+        for step in range(tt_steps):
+            # Prolong
+            V_plus = adj_sol_plus[step].function_space()
+            fwd_sol_plg = Function(V_plus)
+            tm.prolong(out["forward"][step], fwd_sol_plg)
+            adj_sol_plg = Function(V_plus)
+            tm.prolong(out["adjoint"][step], adj_sol_plg)
+
+            # Subtract prolonged adjoint solution from enriched version
+            adj_error = Function(V_plus)
+            adj_sols_plus = split_into_components(adj_sol_plus[step])
+            adj_sols_plg = split_into_components(adj_sol_plg)
+            for i, err in enumerate(split_into_components(adj_error)):
+                err += adj_sols_plus[i] - adj_sols_plg[i]
+
+            # Evaluate errors
+            dwr += dwr_indicator(config, mesh, fwd_sol_plg, adj_error)
+            dwr_list.append(dwr_indicator(config, mesh, fwd_sol_plg, adj_error))
+
+        out["dwr"] = dwr_list
+        
+    out["times"]["estimator"] += perf_counter()
+
+    return out if retall else out["dwr"]
+
+
+def dwr_indicator(config, mesh, q, q_star):
```

Comment:
Again, if this is unmodified then you can just import it from `solving.py`, to avoid duplication.